tests/unit/test_movies.py

```diff
@@ -0,0 +1,617 @@
+                    "popularity": 139.421,
+                    "poster_path": "/q6y0Go1tsGEsmtFryDOJo3dEmqu.jpg",
+                    "release_date": "1994-09-23",
+                    "title": "The Shawshank Redemption",
+                    "vote_average": 8.711,
+                    "vote_count": 25398
+                },
+                {
+                    "backdrop_path": "/rSPw7tgCH9c6NqICZef4kZjFOQ5.jpg",
+                    "genre_ids": [
+                    18,
+                    80
+                    ],
+                    "id": 238,
+                    "original_language": "en",
+                    "original_title": "The Godfather",
+                    "overview": "Spanning the years 1945 to 1955, a chronicle of the fictional Italian-American Corleone crime family. When organized crime family patriarch, Vito Corleone barely survives an attempt on his life, his youngest son, Michael steps in to take care of the would-be killers, launching a campaign of bloody revenge.",
+                    "popularity": 129.69,
+                    "poster_path": "/3bhkrj58Vtu7enYsRolD1fZdja1.jpg",
+                    "release_date": "1972-03-24",
+                    "title": "The Godfather",
+                    "vote_average": 8.7,
+                    "vote_count": 19340
+                },
+                {
+                    "backdrop_path": "/kGzFbGhp99zva6oZODW5atUtnqi.jpg",
+                    "genre_ids": [
+                    18,
+                    80
+                    ],
+                    "id": 240,
+                    "original_language": "en",
+                    "original_title": "The Godfather Part II",
+                    "overview": "In the continuing saga of the Corleone crime family, a young Vito Corleone grows up in Sicily and in 1910s New York. In the 1950s, Michael Corleone attempts to expand the family business into Las Vegas, Hollywood and Cuba.",
+                    "popularity": 71.736,
+                    "poster_path": "/hek3koDUyRQk7FIhPXsa6mT2Zc3.jpg",
+                    "release_date": "1974-12-20",
+                    "title": "The Godfather Part II",
+                    "vote_average": 8.6,
+                    "vote_count": 11667
+                }
+            ]
+        }
+        )
+        expected = {
+            "movies": [
+                    {
+                    "id": 278,
+                    "title": "The Shawshank Redemption"
+                    },
+                    {
+                    "id": 238,
+                    "title": "The Godfather"
+                    },
+                    {
+                    "id": 240,
+                    "title": "The Godfather Part II"
+                    }
+            ]
+        }
+        data = src.movies.toprated_movies(1, headers)
+        actual = src.movies.extract_movie_data(data)
+        assert actual == expected
```

Comment:
This one is a bit hard to explain/is more based off instinct, so I'm gonna take a stab at it and you tell me what parts aren't as clear/I'm not explaining well enough.

---
There's almost two (kind of related?) things going on here:

1. ```python
    # copied the same testing style as popular movies refer to that for more detailed comments
    ```

    This is a good instinct (being consistent for others), and works well for the other test cases where it's the same pattern (the function you're testing is making a request, doing something with the response, and returning that).

    But here, the function you're testing (based on the test name) is `extract_movie_data()`. Since that function doesn't make a request, there *shouldn't* be any need for us to mock a request, which leads into the second thing.

2. We *do need* to mock a request because `toprated_movies()` makes a request, and we're calling that function in our unit test. *Generally*[^1], we don't want to call other unrelated functions in a unit test for a given function. Here, we already know the exact response that `toprated_movies()` is going to give back/know what we want to pass to `extract_movie_data()` (it's exactly what we defined in the `json` field of the mock), so there isn't any need for us to use `toprated_movies()`.

[^1]: there are very rare exceptions to this rule, like when there is some setup being done by our app that our function (that we're testing) relies on. Even then, there are other mechanisms to mock side-effects like that (they're called fixtures in pytest). If your functions/application are designed/implemented correctly, you will *never* run into a scenario where you can't unit test properly.

Reply:
I do understand it because I realized on test_actors that I was doing too much and I was too lazy to change it. I will change it though since it simplifies the testing.